modules/programs/hexchat.nix

```diff
@@ -0,0 +1,248 @@
+      type = nullOr package;
+    };
+  };
+
+  config = mkIf cfg.enable {
+    home.packages = [ pkgs.hexchat ];
+    xdg.configFile."hexchat" = mkIf (cfg.theme != null) {
+      source = cfg.theme.out;
+      recursive = true;
+    };
+    xdg.configFile."hexchat/hexchat.conf" = mkIf (cfg.settings != null) {
+      force = true;
+      text = let hexchatConf = cfg.settings;
+      in concatMapStringsSep "\n" (x: x + " = " + hexchatConf.${x})
+      (attrNames hexchatConf);
+    };
+    xdg.configFile."hexchat/servlist.conf" = mkIf (cfg.channels != null)
+      (if attrNames cfg.channels == [ ] then {
+        text = "";
+      } else {
+        force = true;
+        text = let
+          transformChannel = (channelName:
+            let
+              channel = cfg.channels.${channelName};
+              transformField = (k: v:
+                if v == "" then
+                  ""
+                else ''
+
+                  ${k}=${v}'');
+              listChar = c: l: concatMapStrings (transformField c) l;
+              loginMethodMap = {
+                nickServMsg = 1;
+                nickServ = 2;
+                challengeAuth = 4;
+                sasl = 6;
+                serverPassword = 7;
+                saslExternal = 10;
+                customCommands = 9;
+              };
+            in let
+              name = transformField "N" channelName;
+              nickname = transformField "I" channel.nickname;
+              nickname2 = transformField "i" channel.nickname2;
+              realName = transformField "R" channel.realName;
+              userName = transformField "U" channel.userName;
+              password = transformField "P" channel.password;
+              charset = transformField "E" channel.charset;
+              loginMethod = transformField "L"
+                (toString loginMethodMap.${channel.loginMethod});
+              servers = listChar "S" channel.servers;
+              autojoin = listChar "J" channel.autojoin;
+              commands = listChar "C" channel.commands;
+              options = let
+                computeFieldsValue = (with channel.options;
+                  (if channel.options == null then
+                    0
+                  else
+                    ((if autoconnect then 8 else 0)
+                      + (if !connectToSelectedServerOnly then 1 else 0)
+                      + (if !bypassProxy then 16 else 0)
+                      + (if useSSLForAllServers then 4 else 0)
+                      + (if acceptInvalidSSLCertificates then 32 else 0)
+                      + (if useGlobalUserInformation then 2 else 0))));
+              in transformField "F" (toString computeFieldsValue);
+              # Note: Missing option `D=`.
+            in name + loginMethod + charset + options + # `D=` position +
+            nickname + nickname2 + realName + userName + password + servers
+            + autojoin + commands);
```

Comment:
I still think this code is highly difficult to understand. I would really like to see less usage of let here.

But like I said above, this is mostly IMO, so if you want to keep it like this go on.

Reply:
If you want I can try to show how I would do a rewrite of this part.

Reply:
Which part specifically you find highly difficult to understand?

Reply:
Too many lets make the code difficult to understand. I program in Clojure, they also have lets, even them it is not common to you define let inside lets (sometimes it does, but never more than two).

Reply:
Also, many things don't need to be defined right here, and this makes the code distracting.